contrib/NeonVM/string.c

```diff
@@ -121,26 +146,90 @@ TString *string_appendCString(TString *s, const char *ns)
 {
     s->data = realloc(s->data, s->length + strlen(ns));
     if (!s->data) {
-        fatal_error("Could not allocate appended string data.");
+        fatal_error("Could not allocate appended C string data.");
     }
 
     memcpy(&s->data[s->length], ns, strlen(ns));
     s->length += strlen(ns);
     return s;
+}
+
+void string_resizeString(TString *s, size_t n)
+{
+    s->data = realloc(s->data, n);
+    if (!s->data) {
+        fatal_error("Could not resize string to new count of %ld", n);
+    }
+    s->length = n;
 }
 
 TString *string_appendString(TString *s, TString *ns)
 {
     s->data = realloc(s->data, s->length + ns->length);
     if (!s->data) {
-        fatal_error("Could not allocate appended string data.");
+        fatal_error("Could not allocate appended TString data.");
     }
 
     memcpy(&s->data[s->length], ns->data, ns->length);
     s->length += ns->length;
     return s;
 }
 
+/* NOTE: The caller is responsible for freeing the string created from this function! */
+char *string_asCString(TString *s)
+{
+    char *r = malloc(s->length + 1);
+    /* ToDo: Walk buffer, escape any NUL's that might exist? */
+    memcpy(r, s->data, s->length);
+    r[s->length] = '\0'; /* NUL terminate the string. */
+    return r;
+}
+
+BOOL string_startsWith(TString *self, TString *s)
+{
+    size_t i;
+
+    for (i = 0; i < self->length, i < s->length; i++) {
+        if (self->data[i] != s->data[i]) {
+            return FALSE;
+        }
+    }
+
+    return TRUE;
+}
+
+char *tprintf(char *dest, TString *s)
+{
+    if (!s) {
+        return "";
+    }
+    char *d = malloc(s->length + 1);
+    memcpy(d, s->data, s->length);
+    d[s->length] = '\0';
+    snprintf(dest, TSTR_N, "%s", d);
+    free(d);
+    return dest;
+}
+
+
+#ifdef __STRING_TESTS
+int main()
+{
+    TString *s1 = string_createCString("This is String #1's Data.");
+    TString *s2 = string_createCString("This is String #2's Data.");
+    TString *foo = string_createCString("foo");
+    TString *bar = string_createCString("bar");
+    TString *foobar = string_createCString("foo.bar");
+
+    if (string_startsWith(foobar, foo)) {
+        printf("%s starts with %s\n", string_asCString(foobar), string_asCString(foo));
+    }
+    if (string_startsWith(foobar, bar)) {
+        printf("%s starts with %s\n", string_asCString(foobar), string_asCString(foo));
+    }
+    return 0;
+}
+#endif
 #ifdef _MSC_VER
```

Comment:
It's unnecessary to pop warnings at the end of the file.

Reply:
Will remove this in a near-future commit.